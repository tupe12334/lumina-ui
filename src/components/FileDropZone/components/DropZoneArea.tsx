import { DropZoneContent } from './DropZoneContent'
import { cn } from '../infrastructure/utils'
import styles from '../FileDropZone.module.css'

interface DropZoneAreaProps {
  icon?: string
  title: string
  description: string
  buttonText: string
  disabled: boolean
  isDragActive: boolean
  maxFiles: number
  maxFileSize: number
  className?: string
  onDragEnter?: (e: React.DragEvent<HTMLDivElement>) => void
  onDragOver?: (e: React.DragEvent<HTMLDivElement>) => void
  onDragLeave?: (e: React.DragEvent<HTMLDivElement>) => void
  onDrop?: (e: React.DragEvent<HTMLDivElement>) => void
  onClick?: () => void
  onKeyDown?: (e: React.KeyboardEvent<HTMLDivElement>) => void
  onButtonClick: (e: React.MouseEvent<HTMLButtonElement>) => void
  children: React.ReactNode
}

export function DropZoneArea({
  icon,
  title,
  description,
  buttonText,
  disabled,
  isDragActive,
  maxFiles,
  maxFileSize,
  className,
  onDragEnter,
  onDragOver,
  onDragLeave,
  onDrop,
  onClick,
  onKeyDown,
  onButtonClick,
  children
}: DropZoneAreaProps) {
  const dropZoneClasses = cn(
    styles.dropzone,
    {
      [styles.dragActive]: isDragActive && !disabled,
      [styles.disabled]: disabled
    },
    className
  )

  return (
    <div
      className={dropZoneClasses}
      onDragEnter={disabled ? undefined : onDragEnter}
      onDragOver={disabled ? undefined : onDragOver}
      onDragLeave={disabled ? undefined : onDragLeave}
      onDrop={disabled ? undefined : onDrop}
      onClick={disabled ? undefined : onClick}
      role="button"
      tabIndex={disabled ? -1 : 0}
      aria-label="File upload area"
      aria-disabled={disabled}
      onKeyDown={onKeyDown}
    >
      <DropZoneContent
        icon={icon}
        title={title}
        description={description}
        buttonText={buttonText}
        disabled={disabled}
        maxFiles={maxFiles}
        maxFileSize={maxFileSize}
        onButtonClick={onButtonClick}
      />

      {children}
    </div>
  )
}