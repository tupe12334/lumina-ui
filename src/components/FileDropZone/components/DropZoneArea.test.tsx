import { render, screen } from '@testing-library/react'
import { vi } from 'vitest'
import { DropZoneArea } from './DropZoneArea'

const defaultProps = {
  isDragActive: false,
  disabled: false,
  onDragEnter: vi.fn(),
  onDragLeave: vi.fn(),
  onDragOver: vi.fn(),
  onDrop: vi.fn(),
  onClick: vi.fn(),
}

describe('DropZoneArea', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  test('renders correctly with default props', () => {
    const { container } = render(
      <DropZoneArea {...defaultProps}>
        <div className="test-content">Drop zone content</div>
      </DropZoneArea>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders children content', () => {
    render(
      <DropZoneArea {...defaultProps}>
        <div className="test-content">Test content</div>
      </DropZoneArea>
    )

    expect(screen.getByText('Test content')).toBeInTheDocument()
  })

  test('has correct accessibility attributes', () => {
    render(
      <DropZoneArea {...defaultProps}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button', { name: 'File upload area' })
    expect(dropZone).toHaveAttribute('tabIndex', '0')
    expect(dropZone).toHaveAttribute('aria-label', 'File upload area')
    expect(dropZone).toHaveAttribute('aria-disabled', 'false')
  })

  test('shows disabled state correctly', () => {
    render(
      <DropZoneArea {...defaultProps} disabled={true}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')
    expect(dropZone).toHaveAttribute('aria-disabled', 'true')
    expect(dropZone).toHaveClass('disabled')
  })

  test('shows active drag state correctly', () => {
    render(
      <DropZoneArea {...defaultProps} isDragActive={true}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')
    expect(dropZone).toHaveClass('dragActive')
  })

  test('calls onClick when clicked and not disabled', () => {
    render(
      <DropZoneArea {...defaultProps}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')
    dropZone.click()

    expect(defaultProps.onClick).toHaveBeenCalled()
  })

  test('does not call onClick when disabled', () => {
    render(
      <DropZoneArea {...defaultProps} disabled={true}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')
    dropZone.click()

    expect(defaultProps.onClick).not.toHaveBeenCalled()
  })

  test('applies custom className', () => {
    render(
      <DropZoneArea {...defaultProps} className="custom-class">
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')
    expect(dropZone).toHaveClass('custom-class')
  })

  test('handles keyboard events correctly', () => {
    render(
      <DropZoneArea {...defaultProps}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')

    // Test Enter key
    dropZone.focus()
    dropZone.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }))
    expect(defaultProps.onClick).toHaveBeenCalled()

    vi.clearAllMocks()

    // Test Space key
    dropZone.dispatchEvent(new KeyboardEvent('keydown', { key: ' ' }))
    expect(defaultProps.onClick).toHaveBeenCalled()
  })

  test('prevents keyboard events when disabled', () => {
    render(
      <DropZoneArea {...defaultProps} disabled={true}>
        <div className="test-content">Content</div>
      </DropZoneArea>
    )

    const dropZone = screen.getByRole('button')

    dropZone.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }))
    expect(defaultProps.onClick).not.toHaveBeenCalled()
  })
})