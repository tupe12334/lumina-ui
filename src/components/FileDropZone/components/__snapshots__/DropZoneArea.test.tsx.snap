// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropZoneArea > renders correctly with default props 1`] = `
<div
  aria-disabled="false"
  aria-label="File upload area"
  class="_dropzone_3af3de"
  role="button"
  tabindex="0"
>
  <div
    class="_content_3af3de"
  >
    <svg
      class="_icon_3af3de"
      fill="none"
      height="48"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="48"
    >
      <path
        class="_icon_3af3de"
        d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
      />
      <polyline
        class="_icon_3af3de"
        points="7,10 12,5 17,10"
      />
      <line
        class="_icon_3af3de"
        x1="12"
        x2="12"
        y1="5"
        y2="15"
      />
    </svg>
    <h3
      class="_title_3af3de"
    />
    <p
      class="_description_3af3de"
    />
    <button
      aria-label="Browse files"
      class="_button_3af3de"
      type="button"
    />
    <p
      class="_description_3af3de"
    >
      Max 
       files â€¢ Max 
      MB per file
    </p>
  </div>
  <div
    class="test-content"
  >
    Drop zone content
  </div>
</div>
`;
